ruma-events-macros/src/event_content.rs

```diff
@@ -7,39 +7,164 @@ use syn::{
+    /// redacted.
+    SkipRedacted,
+
+    /// This attribute signals that the events redacted form is manually implemented and should
+    /// not be generated.
+    RedactedCustom,
+}
+
+impl EventMeta {
+    fn get_event_type(&self) -> Option<&LitStr> {
+        if let Self::Type(lit) = self {
+            Some(lit)
+        } else {
+            None
+        }
+    }
 }
 
 impl Parse for EventMeta {
     fn parse(input: ParseStream) -> syn::Result<Self> {
-        input.parse::<Token![type]>()?;
-        input.parse::<Token![=]>()?;
-        Ok(EventMeta::Type(input.parse::<LitStr>()?))
+        if input.parse::<Token![type]>().is_ok() {
+            input.parse::<Token![=]>()?;
+            Ok(EventMeta::Type(input.parse::<LitStr>()?))
+        } else if input.parse::<kw::skip_redaction>().is_ok() {
+            Ok(EventMeta::SkipRedacted)
+        } else if input.parse::<kw::custom_redacted>().is_ok() {
+            Ok(EventMeta::RedactedCustom)
+        } else {
+            Err(syn::Error::new(input.span(), "not a `ruma_event` attribute"))
+        }
     }
 }
 
 /// Create an `EventContent` implementation for a struct.
-pub fn expand_event_content(input: DeriveInput) -> syn::Result<TokenStream> {
+pub fn expand_event_content(input: &DeriveInput) -> syn::Result<TokenStream> {
     let ident = &input.ident;
 
-    let event_type_attr =
-        input.attrs.iter().find(|attr| attr.path.is_ident("ruma_event")).ok_or_else(|| {
-            let msg = "no event type attribute found, \
+    let content_attr = input
+        .attrs
+        .iter()
+        .filter(|attr| attr.path.is_ident("ruma_event"))
+        .map(|attr| attr.parse_args::<EventMeta>())
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let event_type = content_attr.iter().find_map(|a| a.get_event_type()).ok_or_else(|| {
+        let msg = "no event type attribute found, \
             add `#[ruma_event(type = \"any.room.event\")]` \
             below the event content derive";
 
-            syn::Error::new(Span::call_site(), msg)
-        })?;
+        syn::Error::new(Span::call_site(), msg)
+    })?;
+
+    let redacted = if needs_redacted(&input) {
+        let doc = format!("The payload for a redacted {}", ident);
+        let redacted_ident = quote::format_ident!("Redacted{}", ident);
+        let kept_redacted_fields = if let syn::Data::Struct(syn::DataStruct {
+            fields: syn::Fields::Named(syn::FieldsNamed { named, .. }),
+            ..
+        }) = &input.data
+        {
+            let mut fields = named
+                .iter()
+                .filter(|f| {
+                    f.attrs.iter().find_map(|a| a.parse_args::<EventMeta>().ok())
+                        == Some(EventMeta::SkipRedacted)
+                })
+                .cloned()
+                .collect::<Vec<_>>();
+            // remove only the field attributes that are from `ruma_events_macro` compiler
+            // fails otherwise.
```

Comment:
I was briefly unclear on what this does. I think this would be clearer:

```suggestion
            // don't re-emit `ruma_events` attributes
```

WDYT?

Reply:
Much better I was having trouble puting that into words, clearly xD